### ISSUE TITLE
Node and requirejs typings messing up my build

### ISSUE THREAD

**Issue by EricSch** [rollup/rollup#2142]
Since the Node typings are moved to "dependencies" in package.json, they conflict with the requirejs typings `    "@types/requirejs": "2.1.31"`.
I understand, it isn't exactly a rollup issue, but have they to be in "dependencies" and not like before in the "devDependencies" section? 
Thanks for your great work

**Comment by guybedford** [comment]
Can you explain exactly how this type dependency is conflicting in your app? What issues are you seeing here?

**Comment by EricSch** [comment]
After updating to the latetest rollup version, my build suddenly failed while compiling:

`node_modules/@types/node/index.d.ts(5950,5): error TS2300: Duplicate identifier 'mod'.
node_modules/@types/requirejs/index.d.ts(38,11): error TS2300: Duplicate identifier 'mod'.
node_modules/@types/requirejs/index.d.ts(422,13): error TS2403: Subsequent variable declarations must have the same type.  Variable 'require' must be of type 'NodeRequire', but here has type 'Require'.`

I was wondering why the Node typings were installed and I found out, they where included by the rollup package.json. Without the node or requirejs typings, I don't get the error. I'm using Angular 5.0.1 with TS 2.7.2
I'm aware that has nothing do with rollup, but was introduced in our app with the change in your package.json. When you think, I should go to the typings 'maker', I totally understand :ok_hand:

**Comment by guybedford** [comment]
@EricSch it seems like this is a particular aspect of your app that the Node types are not compatible here. I'd suggest adding a custom exclude to your tsconfig.json file to ignore the types - `"exclude": ["node_modules/@types/node"]` should work I think.

We do need to include the Node types as a dependency for our types to work in the first place, so removing isn't really an option unfortunately. For as amazing as TypeScript is it has never quite got JS modularity right :P

**Comment by EricSch** [comment]
@guybedford Thanks for answering, will try that,
I didn't mean to remove them totally from your package, only move them back to the "devDependencies".

And I would say, Typescript has to fight with the JS modularity mess. There are too many different module systems to get everything in every case right :wink:

**Comment by lukastaegert** [comment]
We could also make the `@types/node` a peer dependency, i.e. bring your own types? Though of course then non-TypeScript users will have to deal with a nagging warning that they are missing some dependency they do no care about.

**Comment by lukastaegert** [comment]
If we move them entirely to the dev dependencies, this will leave people with broken types and no information what they have to do to fix that.

**Comment by guybedford** [comment]
Yeah peerDependencies doesn't seem to quite solve it either unfortunately. Marking as a wontfix but happy to discuss further.

**Comment by EricSch** [comment]
@guybedford Excluding the node types doesn't work. TS includes automatically every typings in @types.
One solutions would be to include all the typings separately:
`   "compilerOptions": {
       "types" : ["node", "lodash", "express"]
   }`
But this would be cumbersome. 
I still don't get it why they have to be in the dependencies? My build (via Gulp and Rollup) works fine without the node typings. So when are the typings needed? When using the Rollup API directly?

**Comment by guybedford** [comment]
@EricSch the fact that the "node" type specifically is breaking for you in a Node workflow is a sign that the RequireJS typing approach is just plain wrong. In particular I would argue that this line - https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/requirejs/index.d.ts#L422 is invalid as it is altering an environment invariant, while RequireJS in NodeJS will not actually override the `require` global.

**Comment by EricSch** [comment]
@guybedford I removed the requireJs typings from our package.json. They were obsolete. 
But now with the latest node typings, they are messing up again:
`node_modules/@types/node/index.d.ts(2381,15):   error TS2300: Duplicate identifier 'URL'. node_modules/@types/node/index.d.ts(2399,15):   error TS2300: Duplicate identifier 'URLSearchParams'. node_modules/@types/node/index.d.ts(2417,14):   error TS2661: Cannot export 'URL'. Only local declarations can be exported   from a module. node_modules/@types/node/index.d.ts(2417,19):   error TS2661: Cannot export 'URLSearchParams'. Only local declarations can be   exported from a module.
`

seems an [issue](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/25342) with the typings.

So, those node typings are giving me only troubles. Specially annoying because I don't need them.
Couldn't you have your own typings?

**Comment by guybedford** [comment]
@EricSch you can fix that by locking the broken Node typing to the last working version.

The hard constraint we have here is to ensure Rollup can export working typings that don't give users warnings. And leaving out the Node typing will cause warnings in VSCode and when validating types.

The only alternative I can possibly think of would be to inline our Node typings, specifically inlining the entire definition of EventEmitter in src/rollup/types.d.ts.

**Comment by kumaresan-subramani** [comment]
Hi,
i am also having same problem,
please provide any workaround for this?
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31628

**Comment by bard** [comment]
Having node types in rollup leads to another nasty issue: it makes code compile that shouldn't.

For example, with this `tsconfig.json` I'm telling TypeScript that it should stick to ES5, because I'm targeting environments that don't support any further spec:

```
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["es5"],
...

And indeed, if I forget from what spec `find()` comes from and write somewhere in my sources:

```js
const a = ['foo', 'bar']
console.log(a.find(s => s.indexOf('oo') !== -1))
```

tsc helpfully complains:

```
$ tsc
src/index.ts:6:15 - error TS2339: Property 'find' does not exist on type 'string[]'.

6 console.log(a.find(s => s.indexOf('oo') !== -1))
```

Now I know I have to either get rid of it or import a polyfill.

However, if I want to bundle this project with rollup, due to @types/node being installed I get:

```
$ tsc
Done in 10.26s.
$
```

I.e. code compiles fine now and breaks at runtime.

Ideas on how I can address this?

**Comment by bard** [comment]
In case anyone else stumbles on this: found workaround in https://github.com/Microsoft/TypeScript/issues/18588#issuecomment-432247126 via https://github.com/Microsoft/TypeScript/issues/17042.

In package.json:

```
"postinstall": "shx rm -rf node_modules/@types/node && echo 'workaround for libs importing @types/node on browser environment'"
```

**Comment by shellscape** [comment]
Hey folks. This is a saved-form message, but rest assured we mean every word. The Rollup team is attempting to clean up the Issues backlog in the hopes that the active and still-needed, still-relevant issues bubble up to the surface. With that, we're closing issues that have been open for an eon or two, and have gone stale like pirate hard-tack without activity.

We really appreciate the folks have taken the time to open and comment on this issue. Please don't confuse this closure with us not caring or dismissing your issue, feature request, discussion, or report. The issue will still be here, just in a closed state. If the issue pertains to a bug, please re-test for the bug on the latest version of Rollup and if present, please tag @shellscape and request a re-open, and we'll be happy to oblige.

**Comment by lukastaegert** [comment]
Sorry for the inconvenience, Rollup@2 will finally fix this: #3395

**Comment by lukastaegert** [comment]
Solved in rollup@2.0.0

**Comment by 8Observer8** [comment]
I solved this problem by deleting "requirejs" from "tsconfig.json" > "compilerOptions" > "types".

And I add "node" to "types":

```
        "types": [
            "node"
        ]
```
